server/cluster/cluster.go

```diff
@@ -914,9 +922,140 @@ func (c *RaftCluster) PutStore(store *metapb.Store) error {
+			zap.Int("target region count", targetStore.GetRegionCount()))
+		select {
+		case <-time.After(progressMetricsTimer * time.Second):
+			currentProgress := 100.00 * regionCount / targetStore.GetRegionCount()
+			if currentProgress > progress {
+				progress = currentProgress
+				retryTime = 0
+				storeProgressGauge.WithLabelValues(src.Address, storeId, "up").Set(float64(progress))
+			} else {
+				retryTime = retryTime + 1
+				if retryTime > maxRetryLimit {
+					if similar := pickSimilarStore(rc.GetStores(), src); similar != nil &&
+						similar.GetID() != targetStore.GetID() {
+						targetStore = similar
+						continue
+					}
+					storeProgressGauge.WithLabelValues(src.Address, storeId, "up").Set(100.00)
+					log.Warn("store has reach max retry limit", zap.Uint64("store-id", src.GetId()))
+					return progressExceedRetryLimit
+				}
+			}
+			// pick again if target offline
+			if targetStore = rc.GetStore(targetStore.GetID()); targetStore == nil || !targetStore.IsUp() {
+				targetStore = pickSimilarStore(rc.GetStores(), src)
+			}
+		}
+	}
+}
+
+// recordOfflineProgress record the progress of the offline store
+func recordOfflineProgress(storeID uint64, maxRetryTimes int, rc *RaftCluster) storeProgressStatus {
+	store := rc.GetStore(storeID)
+	if store == nil {
+		log.Warn("store not find", zap.Uint64("store-id", storeID))
+		return progressStoreOffLine
+	}
+	storeLabel := strconv.FormatUint(storeID, 10)
+	count := store.GetRegionCount()
+	if count == 0 {
+		storeProgressGauge.WithLabelValues(store.GetAddress(), storeLabel, "offline").Set(100.00)
+		return progressFinish
+	}
+	progress := 0
+	retryTime := 0
+	for {
+		if progress > 95.00 {
+			log.Info("store down finish", zap.Uint64("store-id ", storeID))
+			storeProgressGauge.WithLabelValues(store.GetAddress(), storeLabel, "offline").Set(100.00)
+			return progressFinish
+		}
+		select {
+		case <-time.After(progressMetricsTimer * time.Second):
+			currentProgress := 100.00 * (count - store.GetRegionCount()) / count
+			if currentProgress > progress {
+				progress = currentProgress
+				storeProgressGauge.WithLabelValues(store.GetAddress(), storeLabel, "offline").Set(float64(progress))
+				retryTime = 0
+			} else {
+				retryTime = retryTime + 1
+				if retryTime > maxRetryTimes {
+					log.Warn("store region count not change, it has reach the max retry time limit",
+						zap.Uint64("store-id", storeID))
+					return progressExceedRetryLimit
+				}
+			}
+		}
+		store = rc.GetStore(storeID)
+		if store == nil {
+			log.Warn("store not find", zap.Uint64("store-id", storeID))
+			storeProgressGauge.WithLabelValues(store.GetAddress(), storeLabel, "offline").Set(100.00)
+			return progressStoreOffLine
+		}
+	}
+}
+
+// pickSimilarStore find one target store
+func pickSimilarStore(stores []*core.StoreInfo, src *metapb.Store) (store *core.StoreInfo) {
+	labels := src.GetLabels()
+	var zone string
+	var sack string
+	for _, v := range labels {
+		if v.Key == "zone" {
```

Comment:
You can consider fetching relevant data from the `location-labels` of config, instead of hard-coding the `key`. In addition, you need to consider the case where there are no `location-labels`.

Reply:
Do it pick without lable restriction if no location-labels??